File: frontend/src/api/notifications.ts
import api from './client';

export interface NotificationTemplate {
  id: number;
  name: string;
  variable_automation_id?: number | null;
  channel: string;
  subject: string | null;
  body: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface NotificationRule {
  id: number;
  automation_id: number;
  automation_name?: string | null;
  is_active: boolean;
  send_condition?: string | null;
  recipient_type: string;
  preferred_channel: string;
  fallback_channel?: string | null;
  template_id?: number | null;
  template_name?: string | null;
  requires_approval: boolean;
  notify_manager: boolean;
  manager_condition?: string | null;
  params_json: Record<string, any>;
  created_at: string;
  updated_at: string;
}

export interface NotificationHistory {
  id: number;
  execution_id?: number | null;
  automation_id?: number | null;
  automation_name?: string | null;
  employee_id?: number | null;
  employee_name?: string | null;
  channel: string;
  recipient?: string | null;
  subject?: string | null;
  message?: string | null;
  status: string;
  data_envio?: string | null;
  sent_at?: string | null;
  error?: string | null;
  attempts: number;
  simulation: boolean;
  created_at: string;
}

export interface NotificationTemplateVariables {
  automation_id: number;
  source_run_ids: number[];
  variables: string[];
  samples: Record<string, any>;
  aliases: Record<string, string>;
}

export const listNotificationTemplates = async () =>
  (await api.get<NotificationTemplate[]>('/notification-templates')).data;

export const listNotificationTemplateVariables = async (automationId: number) =>
  (await api.get<NotificationTemplateVariables>('/notification-template-variables', { params: { automation_id: automationId } })).data;

export const createNotificationTemplate = async (payload: Omit<NotificationTemplate, 'id' | 'created_at' | 'updated_at'>) =>
  (await api.post<NotificationTemplate>('/notification-templates', payload)).data;

export const updateNotificationTemplate = async (id: number, payload: Partial<NotificationTemplate>) =>
  (await api.put<NotificationTemplate>(`/notification-templates/${id}`, payload)).data;

export const deleteNotificationTemplate = async (id: number) => {
  await api.delete(`/notification-templates/${id}`);
};

export const listNotificationRules = async (params?: { automation_id?: number }) =>
  (await api.get<NotificationRule[]>('/notification-rules', { params })).data;

export const createNotificationRule = async (payload: Omit<NotificationRule, 'id' | 'automation_name' | 'template_name' | 'created_at' | 'updated_at'>) =>
  (await api.post<NotificationRule>('/notification-rules', payload)).data;

export const updateNotificationRule = async (id: number, payload: Partial<NotificationRule>) =>
  (await api.put<NotificationRule>(`/notification-rules/${id}`, payload)).data;

export const deleteNotificationRule = async (id: number) => {
  await api.delete(`/notification-rules/${id}`);
};

export const listNotifications = async (params?: { execution_id?: number; automation_id?: number; status?: string; limit?: number; offset?: number }) =>
  (await api.get<NotificationHistory[]>('/notifications', { params })).data;

export const retryNotification = async (id: number) =>
  (await api.post(`/notifications/${id}/retry`)).data;

export const approveNotification = async (id: number) =>
  (await api.post(`/notifications/${id}/approve`)).data;

export const cancelNotification = async (id: number) =>
  (await api.post(`/notifications/${id}/cancel`)).data;

File: backend/app/schemas/notifications.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class NotificationTemplateBase(BaseModel):
    name: str = Field(min_length=2, max_length=200)
    variable_automation_id: int | None = None
    channel: str = "email"
    subject: str | None = None
    body: str = Field(min_length=1)
    is_active: bool = True


class NotificationTemplateCreate(NotificationTemplateBase):
    pass


class NotificationTemplateUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=2, max_length=200)
    variable_automation_id: int | None = None
    channel: str | None = None
    subject: str | None = None
    body: str | None = Field(default=None, min_length=1)
    is_active: bool | None = None


class NotificationTemplateOut(NotificationTemplateBase):
    id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class NotificationRuleBase(BaseModel):
    automation_id: int
    is_active: bool = True
    send_condition: str | None = None
    recipient_type: str = "responsavel"
    preferred_channel: str = "email"
    fallback_channel: str | None = None
    template_id: int | None = None
    requires_approval: bool = False
    notify_manager: bool = False
    manager_condition: str | None = None
    params_json: dict[str, Any] = Field(default_factory=dict)


class NotificationRuleCreate(NotificationRuleBase):
    pass


class NotificationRuleUpdate(BaseModel):
    automation_id: int | None = None
    is_active: bool | None = None
    send_condition: str | None = None
    recipient_type: str | None = None
    preferred_channel: str | None = None
    fallback_channel: str | None = None
    template_id: int | None = None
    requires_approval: bool | None = None
    notify_manager: bool | None = None
    manager_condition: str | None = None
    params_json: dict[str, Any] | None = None


class NotificationRuleOut(NotificationRuleBase):
    id: int
    template_name: str | None = None
    automation_name: str | None = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class NotificationOut(BaseModel):
    id: int
    execution_id: int | None = None
    automation_id: int | None = None
    automation_name: str | None = None
    employee_id: int | None = None
    employee_name: str | None = None
    channel: str
    recipient: str | None = None
    subject: str | None = None
    message: str | None = None
    status: str
    data_envio: datetime | None = None
    sent_at: datetime | None = None
    error: str | None = None
    attempts: int
    simulation: bool
    created_at: datetime

    class Config:
        from_attributes = True


class NotificationRetryOut(BaseModel):
    id: int
    status: str
    error: str | None = None
    attempts: int
